building-and-packaging/material/exercise_cmake_text.md

```diff
@@ -0,0 +1,27 @@
+
+To get an independent and reproducible environment as common ground, we use and, thus repeat, Docker.
+
+Deadline: **Thursday, December 2nd, 2021, 9am**
+
+## Overview
+
+- Your overall goal is to open a pull request from a fork to [this repository](https://github.com/Simulation-Software-Engineering/cmake-exercise-test/) (TODO: update link) adding a `Dockerfile` and a `CMakeLists.txt`, besides minor changes in `main.cpp` uncommenting some code parts. Please name your pull request `<GitLab-USERNAME> Add building and container recipes`. With your pull request, building and running the docker container, one should be able to directly build the code skeleton within the container using CMake. As many of the currently commented-out dependencies as possible should be supported.
+
+## First steps
+
+- For and clone the repository, have a look at the `main.cpp` and the README.
+- Then, build `main.cpp` manually (e.g. `g++ main.cpp -o main`) and run the program (`./main`).
+
+## Tips / remarks
+
+- **Docker**: As by now, you already know how to set up a basic Docker container, we do no longer provide detailed instructions. Please note the guidelines in the README. We recommend to build up the image incrementally. Start with a rather empty one, install dependencies manually in the interactive mode, and take notes of the commands your use. Once, you know what you want to do, add the commands to the Dockerfile and rebuild your image.
+- **Dependencies**: Also add dependencies one by one: Uncomment the necessary part of `main.cpp`, install the dependency, write the CMake part, build and run. If everything works, go to the next dependency. Maybe start with the boost dependencies. `boost container` is a header-only dependency, `boost filesystem` needs to be linked.
+- **CMake**:
+    - Please follow good practice and create a `build` directory at the root level from which you call `cmake ..`. There is a hard-coded path `../yaml/config.yml` in `yaml/yaml.hpp`. If you use a non-standard location for your `build` directory, you have to adapt this path.
+    - deal.II uses some fancy [CMake macros](https://www.dealii.org/current/users/cmake_user.html).
+    - There is a CMake module to [find boost libraries](https://cmake.org/cmake/help/latest/module/FindBoost.html).
+- **Bonus**: `yaml-cpp` is the optional bonus task. Do this one last. We want to use the newest version `0.7.0`, which is not available through Aptitude. To get it directly from GitHub, you can use `wget` (or `git clone`). Afterwards, build and install (`make install`) yourself. Do not forget to add the necessary instructions to your Dockerfile.
```

Comment:
Another think: I did not find any hint whether one should compile `yaml-cpp` as shared or static library.

Reply:
> Another think: I did not find any hint whether one should compile yaml-cpp as shared or static library.

That's up to them.

> How do you intend students to add this dependency? Using find_package (what I did) or another way?.

After installation, it worked smoothly for me.

> That is why added an additional bash build helper that has this command line hardcoded to my solution.

Students could explain such things also in the PR description.

Reply:
> After installation, it worked smoothly for me.

As we discussed this earlier. I think the problem was that I provided an installation prefix that was not `/usr`. Since you installed everything in the standard path, the linker knows about the location of your library and the compiler knows about the include files. This is at least something to keep in mind in case the students have problems.

Reply:
As discussed yesterday, I still had to add `ENV LIBRARY_PATH $LIBRARY_PATH:/usr/local/lib/`, which is a strange Docker "feature".